File: daa/structures/src/com/tw/vrc/lists/Stack.java
package com.tw.vrc.lists;

public class Stack<T extends Object> {
    private Node<T> tail;

    public void push(T t) {
        Node<T> node = new Node<T>(t);
        node.next = tail;
        tail = node;
    }

    public T pop() {
        Node<T> toPop = tail;
        if (tail != null) {
            tail = tail.next;
            return toPop.value();
        }
        return null;
    }

}

File: algos/graphs/src/com/vijayrc/algos/stack/Stack.scala
package com.vijayrc.algos.stack

class Node(val value:AnyRef, val next:Node){}

class Stack {
  var head:Node = null

  def push(item:AnyRef):Stack = {
    if(head == null) head = new Node(item,null)
    else head = new Node(item, head)
    this
  }
  def pop():AnyRef = {
    if(head == null) null
    val value = head.value
    head = head.next
    value
  }
  def show(){
    var temp = head
    while(temp != null){
      print(temp.value.toString)
      temp = temp.next
    }
  }
}

